main.py

```diff
@@ -0,0 +1,162 @@
+
+if not _is_valid_uuid(secret_key):
+    raise InvalidNotifyKeyError("API key is not a valid uuid")
+
+
+def create_jwt_token(secret: str, client_id: str) -> str:
+    headers = {"typ": "JWT", "alg": "HS256"}
+    claims = {"iss": client_id, "iat": int(datetime.now(timezone.utc).timestamp())}
+    return jwt.encode(payload=claims, key=secret, headers=headers)
+
+
+template_id_mapping = {
+    ("H", "GB-ENG", "en"): "0c5a4f95-bfa4-4364-9394-8499b4d777d5",
+    ("H", "GB-WLS", "en"): "0c5a4f95-bfa4-4364-9394-8499b4d777d5",
+    ("H", "GB-WLS", "cy"): "755d73d1-0cb6-4f2f-95e9-857a2ad071bb",
+    ("H", "GB-NIR", "en"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("H", "GB-NIR", "ga"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("H", "GB-NIR", "eo"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("I", "GB-ENG", "en"): "71de56dc-f83b-4899-93ab-7fe61e417c2e",
+    ("I", "GB-WLS", "en"): "71de56dc-f83b-4899-93ab-7fe61e417c2e",
+    ("I", "GB-WLS", "cy"): "1001ac43-093d-425c-ac7d-68df5147c603",
+    ("I", "GB-NIR", "en"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("I", "GB-NIR", "ga"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("I", "GB-NIR", "eo"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("C", "GB-ENG", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("C", "GB-WLS", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("C", "GB-WLS", "cy"): "e4a4ebea-fcc8-463b-8686-5b8a7320f089",
+}
+
+data_fields = ("email_address", "display_address", "tx_id", "questionnaire_id")
+
+
+session = Session()
+
+
+def _validate_request(request: Request) -> Tuple[Mapping, str, Mapping]:
+    if not request.method == "POST":
+        # Note that Cloud Functions expect serialized JSON
+        # to correctly log
+        # https://cloud.google.com/functions/docs/monitoring/logging#writing_structured_logs
+        raise InvalidRequestError("Method not allowed", 405)
+
+    if not request.json:
+        raise InvalidRequestError("Missing notification request data", 422)
+
+    data = request.json["payload"]["fulfilmentRequest"]
+
+    log_context = {
+        "tx_id": data.get("tx_id"),
+        "questionnaire_id": data.get("questionnaire_id"),
+    }
+
+    required_keys = ("form_type", "region_code", "language_code")
+    if missing_keys := [key for key in required_keys if not data.get(key)]:
+        msg = f"Missing {', '.join(missing_keys)} identifier(s)"
+        raise InvalidRequestError(msg, 422, log_context)
+
+    try:
+        template_id = template_id_mapping[
+            (data["form_type"], data["region_code"], data["language_code"])
+        ]
+    except KeyError as exc:
+        raise InvalidRequestError("No template id selected", 422, log_context) from exc
+
+    return {key: data.get(key) for key in data_fields}, template_id, log_context
+
+
+# pylint: disable=too-many-return-statements
+def send_email(request: Request) -> Tuple[str, int]:
+    try:
+        data, template_id, log_context = _validate_request(request)
```

Comment:
I would say the validate method is also parsing the request, so might be better named `_parse_request`. I think it's clear it's validating as it can throw an `InvalidRequestError` exception.